### ISSUE TITLE
Default value for :sectnums:

### ISSUE THREAD

**Issue by antonmoiseev** [asciidoctor/asciidoctor#1434]
Is it possible to set the default value for numbering sections? Use case: we have an asciidoc file per chapter, each chapter is processed separately, we'd love to set the default value, so the chapter 3 will use numbering 3, 3.1, 3.1.1 and so on.

**Comment by mojavelinux** [comment]
I think this is a duplicate of #1113. As I mention in that issue, I've implemented a prototype of this idea in the extension lab that you can start using today.

https://github.com/asciidoctor/asciidoctor-extensions-lab/blob/master/lib/sectnumoffset-treeprocessor.rb

**Comment by antonmoiseev** [comment]
#1113 looks like even more advanced feature. It allows using `:section-number:` multiple times per document. I was originally thinking about allowing to initialize sectnums with a default value (e.g. `:sectnums: 2`), but since it's a boolean attribute you would probably want to add a new one. Anyway I'd glad to have this feature in any form.

**Comment by mojavelinux** [comment]
Yes, I agree #1113 is a bit wider in scope, but if we solve this problem for one section title, we'll end up solving it for all. The general idea is that it's possible to set the section start number.

I've thought about trying to reuse the `sectnums` for this purpose, but I don't think the mapping is very natural (and easily misconstrued), so I prefer using a separate attribute.

I was thinking that instead of `sectnumoffset` like in the extension in the lab, we should instead align with the terminology for an ordered list and call it `sectnum-start`. That would even open up the door to start at a sub-level (e.g., `:sectnum-start: 2.1`).

With that said, I think I'll start by updating the extension in the lab to use this new attribute name. (Keep in mind you can start using the extension today).

(In the future, this may even open the door for letters instead of numbers, but that's a separate issue).

**Comment by nathany** [comment]
I am in the same situation. `:sectnum-start: 2.1` sounds perfect.

Not only for sections though. I would like to configure listings, tables and figures to all begin with `2.` in the document for Chapter 2.

**Comment by paulvickers** [comment]
I agree with @nathany. As I have floats numbered by chapter (Table 1.2, Figure 5.3, etc) I've been hard-coding table, figure, etc. numbers into the text when making cross references and the proposed functionality would be great.

**Comment by nathany** [comment]
Honestly, I kind've expected everything to just work when I put a `:chapter: 2` attribute in the document.

Though I'm not opposed to configuring things a little more manually. Indeed, my publisher doesn't use section #'s for 4th level headings, so I would need to disable numbering on those ones in any case.

My main concern is to avoid renumbering things when I swap sections around.

**Comment by mojavelinux** [comment]
@nathany It's already possible to control the listings, figures and table numbers using the respective attributes:

```
:listing-number: 1
:table-number: 1
:figure-number: 1
```

(one less than the number you want it to start with).

These are all independent because there isn't a concept (yet) in Asciidoctor that block numbers are correlated with sections in any way. We are discussing the issue related to #979.

It's almost possible to get the subsection numbering using the caption attributes:

```
:table-caption: Table 2.
```

The problem is, we get a space after "Table 2." before "1" (e.g., "Table 2. 1").

The crux of the issue is that we are really stepping outside of the original intent of these attributes and the only way to truly accommodate this use case is to rethink how this works.

...the reason AsciiDoc doesn't do much to help here is that this is one of those cases when AsciiDoc (Python) assumed that DocBook would handle this problem and therefore didn't really add much control that converters could use directly. We have to pave a new path.

**Comment by nathany** [comment]
@mojavelinux I don't know what :listing-number:, etc. do. These don't appear to be documented, or I haven't found the documentation on them.

Right now my listing captions appear with exactly the text I put.
```adoc
.Functions with custom types: temperature.go
```

To format them as my publisher desires, but still have some auto-numbering working for me, I would want something like this:

```adoc
Listing {chapter}.{listing-number} Functions with custom types: temperature.go
```

The `listing-number` attribute and the rest could be generalized. In Go there is an [iota](http://golang.org/ref/spec#Iota) identifier that increments every time it's referenced. Such a feature would allow

:listing-number: iota
:table-number: iota + 1

**Comment by mojavelinux** [comment]
The `*-number` attributes are sort of the internal counter that increment each time they are used for a caption. See https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/abstract_block.rb#L324-L342

**Comment by nathany** [comment]
Either it doesn't work in the Chrome Extension or I'm just plain doing it wrong.

**Comment by Mogztter** [comment]
> Either it doesn't work in the Chrome Extension
If this is the case, feel free to open an issue here: https://github.com/asciidoctor/asciidoctor-chrome-extension/ and I will look at it

**Comment by stallio** [comment]
Continuing the tangent conversation...

> It's almost possible to get the subsection numbering using the caption attributes:

> :table-caption: Table 2.

If you've set a `:chapter-number:` or similar global attribute for the file, then you can do this:

```
:table-caption: Table {chapter-number}.
```
You'll still get the space in the middle, but this way you can at least reuse the same the same attributes on all your documents.

Another thought: it seems like the `*-caption` attributes are doing double-duty, as they're used both to turn numbering on/off and to set the text that goes before the number (or disable that text). Thus for example if you set `:figure-caption!:` then not only will the figure numbers not be shown, but `{figure-number}` will hold no value.

It would be nice to decouple these functions from each other. Then you could move the move the number inside the caption and put it in any format you want. Or you could even have unnumbered captions.

**Comment by mojavelinux** [comment]
> It would be nice to decouple these functions from each other. 

I completely agree. That gets to the heart of the issue. We don't have enough degrees of freedom here to get the output we really want, just workarounds to get something close. This is part of AsciiDoc's legacy that clearly needs some redesign to fit current requirements. 

We've been talking for some time about not only decoupling the attributes that control the object labels, but also to redesign their names so we're calling a spade a spade. It's mostly covered by #702, which is where I think we should discuss this aspect of the issue.

**Comment by mojavelinux** [comment]
FYI, you can influence the start number for chapters.

```
= Document Title
:chapter-number: -1
:sectnums:
:doctype: book

== Chapter A

content

== Chapter B

content
```

However, this only works for chapter numbers, not section or part numbers.

**Comment by kevinlang** [comment]
Related: #1113
